lib/Optimizer/CodeGen/LowerToQIR.cpp

```diff
@@ -549,59 +542,121 @@ class OneTargetOneParamRewrite : public ConvertOpToLLVMPattern<OP> {
+      return instOp.emitError("unsupported control type for " + instName +
+                              ": " + typeName);
+    }
 
-      FlatSymbolRefAttr instSymbolRef =
+    // If the control is a qubit, we need to pack it into an Array*
+    if (numControls == 1 && type.isa<quake::RefType>()) {
+      FlatSymbolRefAttr packingSymbolRef =
           cudaq::opt::factory::createLLVMFunctionSymbol(
-              qirFunctionName, /*return type=*/LLVM::LLVMVoidType::get(context),
-              std::move(tmpArgTypes), parentModule);
+              cudaq::opt::NVQIRPackSingleQubitInArray,
+              /*return type=*/qubitArrayType, {qubitIndexType}, parentModule);
+
+      // Pack the qubit into the array
+      Value result =
+          rewriter
+              .create<LLVM::CallOp>(loc, qubitArrayType, packingSymbolRef,
+                                    adaptor.getControls().front())
+              .getResult();
+      // The array result is what we want for the function args
+      funcArgs.push_back(result);
+    }
 
-      // Add the target op
-      funcArgs.push_back(adaptor.getTargets().front());
+    // If type is a VeqType, then we're good, just forward to the call op
+    if (numControls == 1 && type.isa<quake::VeqType>()) {
+      if (negatedQubitCtrls)
+        return instOp.emitError("unsupported controlled op " + instName +
+                                " with vector of ctrl qubits");
 
-      // Here we already have and Array*, Qubit*
-      rewriter.replaceOpWithNewOp<LLVM::CallOp>(instOp, TypeRange{},
-                                                instSymbolRef, funcArgs);
-      return success();
+      // Add the control array to the args.
+      funcArgs.push_back(adaptor.getControls().front());
     }
 
-    // If the control is a qubit, we need to pack it into an Array*
-    FlatSymbolRefAttr packingSymbolRef =
-        cudaq::opt::factory::createLLVMFunctionSymbol(
-            cudaq::opt::NVQIRPackSingleQubitInArray,
-            /*return type=*/qubitArrayType, {qubitIndexType}, parentModule);
-    // Pack the qubit into the array
-    Value result =
-        rewriter
-            .create<LLVM::CallOp>(loc, qubitArrayType, packingSymbolRef,
-                                  adaptor.getControls().front())
-            .getResult();
-    // The array result is what we want for the function args
-    funcArgs.push_back(result);
-
     // This is a single target op, add that type
     tmpArgTypes.push_back(qubitIndexType);
 
+    // Get function pointer to ctrl operation
     FlatSymbolRefAttr instSymbolRef =
         cudaq::opt::factory::createLLVMFunctionSymbol(
             qirFunctionName, /*return type=*/LLVM::LLVMVoidType::get(context),
             std::move(tmpArgTypes), parentModule);
 
+    // __quantum__qis__NAME__ctl(double, Array*, Qubit*) Type
+    auto instOpQISFunctionType = LLVM::LLVMFunctionType::get(
+        LLVM::LLVMVoidType::get(context), std::move(tmpArgTypes));
```

Comment:
This looks like the second call to `std::move(tmpArgTypes)`. (The other one is on line 598.) Doesn't the first one place the object in an unspecified state?